smdebug/tensorflow/keras.py

```diff
@@ -1383,3 +1332,80 @@ def record_tensor_value(self, tensor_name, tensor_value):
+
+        TODO: Add support for other profiling at the start of the batch.
+        """
+        self.start = time.time()
+
+        if self._is_not_supported():
+            return
+
+        # When training with native TF2, `self.mode_steps[mode]` isn't incremented until the GradientTape callback
+        # for `tape._push_tape` is called. So in this case, the current step is `self.mode_steps[mode] + 1`.
+        current_step = self.mode_steps[mode]
+        if self.is_profiler_enabled_for_native_training:
+            current_step += 1
+
+        self.profiler_config_parser.load_config()
+        self.profiler_config_parser.handle_step_start_python_profiling(mode, current_step)
+
+    def profiling_end_batch(self, mode=ModeKeys.TRAIN):
+        """
+        Enabling profiler at the end of train batch for native Tf2 training.
+        :param mode: ModeKeys.TRAIN ModeKeys.EVAL ModeKeys.PREDICT
+
+        TODO: Add support for other profiling at the end of the batch.
+        """
+        if self._is_not_supported():
+            return
+
+        self.profiler_config_parser.handle_step_end_python_profiling(mode, self.mode_steps[mode])
+
+    def profiling_end(self):
+        """
+        Stop profiler at the end of training for native TF2.
+
+        TODO: Add support for other profiling at the end of training.
+        """
+        # If the hook is closed twice, the process will hang.
+        if self.is_hook_closed:
+            return
+        self.close()  # Unwrap the tape before closing
+        self.is_hook_closed = True
+        self.profiler_config_parser.stop_post_hook_close_python_profiling()
+        self.is_profiler_enabled_for_native_training = False
+
+    def profiler(self, mode=ModeKeys.TRAIN):
+        """
+        Context manager to be inserted directly into the training script to enable profiling for native TF2 training.
+
+        With respect to Python profiling, for a given step `n`, all code inside the context is profiled as during
+        the step n. All code before the context is profiled as between steps n-1 and n. All code after the context
+        is profiled as between steps n and n+1.
+
+        Example:
+
+        ```
+        for epoch in range(n_epochs):
+            for data, labels in dataset:
+                setup()
+                with hook.profiler():
+                    labels = tf.one_hot(labels, depth=10)
+                    with tf.GradientTape() as tape:
+                        logits = train_step(data, labels)
+                    grads = tape.gradient(logits, model.variables)
+                    opt.apply_gradients(zip(grads, model.variables))
+                cleanup()
+        ```
+        """
+        _ = self.wrap_tape(
+            tf.GradientTape()
+        )  # GradientTape functions must be wrapped to correctly manage current step
+        self.set_mode(mode)
+        self.is_profiler_enabled_for_native_training = True
+        return ProfilerContextManager(self, mode)
```

Comment:
See if you can use a `yield` statement instead of managing the context manager with an object.

Reply:
Done